swarm/src/lib.rs

```diff
@@ -762,74 +751,77 @@ where
                 peer_id,
                 id,
                 endpoint,
-                other_established_connection_ids,
+                connection,
                 concurrent_dial_errors,
                 established_in,
             } => {
                 if self.banned_peers.contains(&peer_id) {
-                    // Mark the connection for the banned peer as banned, thus withholding any
-                    // future events from the connection to the behaviour.
-                    self.banned_peer_connections.insert(id);
                     self.pool.disconnect(peer_id);
```

Comment:
Given that we drop `connection` here, is the call to `pool.disconnect` still needed? There should be no connection to the peer in the pool. The previous `Swarm::ban_peer_id` should have already called `pool.disconnect`.

Also, if I am not mistaken, we would previously gracefully close connections to banned peers via the `pool.disconnect` call. Given that `connection` is now just dropped, this is no longer the case, right?

Reply:
> Given that we drop `connection` here, is the call to `pool.disconnect` still needed? There should be no connection to the peer in the pool. The previous `Swarm::ban_peer_id` should have already called `pool.disconnect`.

Originally I thought that we might have other connections to this peer but I guess that isn't actually possible because we would have disconnected those at the time of banning the peer.

> Also, if I am not mistaken, we would previously gracefully close connections to banned peers via the `pool.disconnect` call. Given that `connection` is now just dropped, this is no longer the case, right?

True. I didn't bother much because this code will go away anyway once we implement it via the new `NetworkBehaviour` callbacks but yeah, it is a change in behaviour so we should probably keep this behaviour.

This made me think though, we should probably gracefully close connections in https://github.com/libp2p/rust-libp2p/pull/3254 that have been rejected by a `NetworkBehaviour`.

Reply:
I've pushed some more commits, let me know what you think!